neevaco/neeva-android
# Welcome Flow: Navigation Refactoring + more Login Support

* [Refactor some navigation logic to live in FirstRunModel](https://github.com/neevaco/neeva-android/commit/fd7d205de3fd97653eaed16eb4744b403241fbe5)
* [Add LoginFlowNavModel to keep back stack shallow](https://github.com/neevaco/neeva-android/commit/976567d53f3317cc61455b8c8255af85cfcff20f)
* [Add support for where app navigation leads after a login](https://github.com/neevaco/neeva-android/commit/57df4269f8257266ae7e81551c858b68b50b18b4)
* [Make onNewIntent() in WelcomeFlowActivity less crash prone](https://github.com/neevaco/neeva-android/pull/1177/commits/6787cb2b393bf4fbd7a8ec7c80b33367d714e382)

How login works in this PR:
* In the very First Run of the app, the login buttons will set the `activityToReturnToAfterLogin` to be `WelcomeFlowActivity` and the `screenToReturnToAfterLogin` to be `SET_DEFAULT_BROWSER` (or `FINISH_WELCOME_FLOW` if the default browser is already set to Neeva before launching FirstRun -- edge case)
* If the login is triggered from Settings or the Browser, the `activityToReturnToAfterLogin` will be `NeevaActivity` and the screen will be respectively the Settings or Browser screen.
* The login will happen in the `SingleTabActivity` which will then start an Intent back to the Neeva app once logged in.
* MainActivity will catch this intent, update the cookie jar and cached cookie with the new login cookie, and then open the specified activity based on `activityToReturnToAfterLogin` stored in `SharedPreferences`. 
* The `NeevaActivity` and `WelcomeFlowActivity` both have functions to process this cookie and open the right `screen` based on the `screenToReturnToAfterLogin` stored in `SharedPreferences`. 

## Checklists

### Pointers to help with review
* https://www.figma.com/file/dylvDPB69JG9U1rzZ3HQ52/Android-Premium?node-id=0%3A1&t=i9BOuz5zR0S5EZxX-0

### How was this tested?

### Manual test cases
Login from Settings (Default Browser not set to Neeva) -> Should show SetDefaultBrowser Screen after login.
Login from Settings (Default Browser set to Neeva + Free plan selected in login) -> Should go straight back to Settings after login. 
Login from Spaces prompt (Will deal with this in a follow-up PR!)
Login from FirstRun (Premium not available) -> Show SetDefaultBrowser Screen (Skip Login)
Login from FirstRun (Premium Available)

### Screenshots and screen recordings
Login from Settings (Default Browser not set to Neeva) -> Should show SetDefaultBrowser Screen after login.

https://user-images.githubusercontent.com/20916043/210892547-1b4c7c67-c95e-4004-bdab-a7597f0f5389.mp4


Login from Settings (Default Browser set to Neeva + Free plan selected in login) -> Should go straight back to Settings after login. 

https://user-images.githubusercontent.com/20916043/210899108-eeba0d07-7dee-4b1a-83f6-2d58d95a134c.mp4



Login from FirstRun (Premium not available) -> Show SetDefaultBrowser Screen (Skip Login)

https://user-images.githubusercontent.com/20916043/210860259-981ab2f1-1568-4179-a859-ad5636542a7c.mp4

Login from FirstRun (Premium Available) -> Show SetDefaultBrowserScreen after login -> Show ZeroQuery


https://user-images.githubusercontent.com/20916043/210899007-c4d767e6-717e-4723-b4d8-745f01666b76.mp4





## Issues addressed
* Link(s) to Github issue(s) (if applicable)
